feat: filter historical transactions

iamacook opened the pull request:
## What it solves
Requesting filtered transactions.

## How this PR fixes it
The chosen filter is now applied to the historical transactions, as well as being appended to the URL (with the added benefit of being able to share).

## Screenshots
![image](https://user-images.githubusercontent.com/20442784/172583485-049ae7c2-445b-4c5d-a814-11e6c8841f96.png)

## How to test it
Filter the transactions and observe the filtered data. The filter should also be appended to the URL. Opening the filter-appended URL should load the historical transactions, filtered.

iamacook commented on src/components/Root/index.tsx:
```diff
@@ -31,6 +33,9 @@ const RootConsumer = (): React.ReactElement | null => {
 
   useEffect(() => {
+    // Initialize the SDK
+    setBaseUrl(GATEWAY_URL)
```
Updating to the new SDK required the `baseUrl` to be set. Alongisde this, passing of `GATEWAY_URL` have been removed app-wide.

katspaugh commented on src/logic/safe/store/actions/transactions/fetchTransactions/index.ts:
```diff
@@ -1,29 +1,37 @@
 import { ThunkDispatch } from 'redux-thunk'
 import { AnyAction } from 'redux'
+import { parse } from 'querystring'
```
Can we use the native URL parser instead?

katspaugh replied:
https://developer.mozilla.org/en-US/docs/Web/API/URL – it has a `searchParams` property

iamacook replied:
I'll pivot to using `URLSearchParams` as I did in `getHistoryTxListPage` instead.

francovenica commented:
Some early feedback while the ticket is still in peer review.

Safe for testing:
https://pr3870--safereact.review-safe.gnosisdev.com/app/rin:0x9913B9180C20C6b0F21B6480c84422F6ebc4B808/transactions/history

Suggestion:
And empty state when nothing is found when filtering. Right now it shows the default image with the message "History transactions will appear here". A "no results found" image for example.

Issue:
1 - The date is not saved after filtering. The inputs return to the default state, but they are still there, since if you try to add another input (like nonce) and apply the filter it will use the dates again.

2 - The from/to filter is not working properly for outgoing tx. In the safe I filtering by outgoing tx, from May 2021 to Dec 2021 and the results still show tx from 2022. The payload only shows the "greater or equal than" parameter, but in the URL all the parameters are there
![image](https://user-images.githubusercontent.com/4503659/169548865-34d111f5-fc58-4cb5-a5ec-17e64180c6d4.png)
A gif to show how I'm filtering. In this case it actually have the "less or equal" in the payload, something that didn't happen before, but the results still show tx from 2022:
![test1](https://user-images.githubusercontent.com/4503659/169552096-559f0588-ace0-4ebd-a9b1-ad5938abd7ec.gif)

3 - Filtering by outgoing tx seems to filter only by those tx that send the main token of the network (ETH in this case)
Try to search in that safe the nonce 1000 
Then try to search by this address as recipient rin:0x6E45d69a383CECa3d54688e833Bd0e1388747e6B
The tx with nonce 1000 is a "Send funds" to that address, but DAI were sent. When searching for that address, you only see "Send funds" of ETH tokens.
![test1](https://user-images.githubusercontent.com/4503659/169554517-8a1e1778-4d7d-42af-97a0-0c7d65fd3845.gif)

iamacook commented:
@franco, thanks for the early testing!

- I've added a 'No results found' message.
- The date fields now use the values in the URL as their default values.
- I've checked the [swagger](https://safe-transaction.rinkeby.staging.gnosisdev.com/) and when directly querying the transaction service, the correct period is returned. However, it isn't [via the gateway](https://safe-client.staging.gnosisdev.com/v1/chains/4/safes/0x9913B9180C20C6b0F21B6480c84422F6ebc4B808/multisig-transactions/?execution_date__gte=2021-05-20T00:00:00.000Z&execution_date__lte=2021-12-31T00:00:00.000Z&executed=true). cc @fmrsabino
- This also happens on the transaction service (after checking the swagger) so I'm not too sure about this one. I'll have to confer with backend on this as well. ([Here's the gateway query](https://safe-client.staging.gnosisdev.com/v1/chains/4/safes/0x9913B9180C20C6b0F21B6480c84422F6ebc4B808/multisig-transactions/?execution_date__gte=2021-05-20T00:00:00.000Z&execution_date__lte=2021-12-31T00:00:00.000Z&executed=true) if we need it).

iamacook commented:
I've removed date filtering for the time being.

katspaugh commented on src/routes/safe/components/Transactions/TxList/Filter/index.tsx:
```diff
@@ -230,7 +246,7 @@ const Filter = (): ReactElement => {
                             type="date"
                             control={control}
-                          />
+                          /> */}
```
Could be disabled + tooltip "Coming soon"

iamacook replied:
What do you think of this? 
![image](https://user-images.githubusercontent.com/20442784/172343884-714772c5-0fd7-4183-b5d6-ff012f488823.png)

iamacook replied:
Tooltip seems a bit odd imo
![image](https://user-images.githubusercontent.com/20442784/172344488-4a57379a-d26b-4a61-8582-dab731874661.png)

liliiaorlenko replied:
![clock](https://user-images.githubusercontent.com/94694401/172345570-d5ecf650-39fb-4dfc-894f-2c8e002d4176.svg)
how about clock icon on the left and on hover of the icon - coming soon?

liliiaorlenko replied:
+ disabled field

liliiaorlenko replied:
and no green border highlight

iamacook replied:
I've implemented the following:
![image](https://user-images.githubusercontent.com/20442784/172363890-524aa6fe-b56c-4ac8-81b1-0566c37dc8b8.png)

iamacook replied:
After going through this with @usame-algan, I've adjusted the logo to an hourglass:
![image](https://user-images.githubusercontent.com/20442784/172584605-f924a978-7ba1-419b-9267-f5ad9b960a33.png)

DiogoSoaress commented:
If you apply the same filter twice I think it moves to the next page 🧐 can you reproduce it?

For instance

<img width="1543" alt="Screenshot 2022-06-07 at 15 40 59" src="https://user-images.githubusercontent.com/32431609/172397233-fa461c27-6478-459a-bf4e-d7f35e09349d.png">

iamacook commented:
> If you apply the same filter twice I think it moves to the next page 🧐 can you reproduce it?
> 
> For instance
> 
> <img alt="Screenshot 2022-06-07 at 15 40 59" width="1543" src="https://user-images.githubusercontent.com/32431609/172397233-fa461c27-6478-459a-bf4e-d7f35e09349d.png">

I've adjusted it to no longer request the same filter twice.

DiogoSoaress commented on src/logic/safe/store/actions/transactions/fetchTransactions/loadGatewayTransactions.ts:
```diff
@@ -1,13 +1,96 @@
+      return pageUrl
+    }
+
+    const url = new URL(pageUrl)
```
Should be inside a `try..catch` as `pageUrl` can be undefined

iamacook replied:
There's a guard a few lines up.

iamacook replied:
After testing, it seems like there are a few values that could throw here. I'll add it.

DiogoSoaress commented on src/logic/safe/store/actions/transactions/fetchTransactions/loadGatewayTransactions.ts:
```diff
@@ -1,13 +1,96 @@
+      url.searchParams.set(key, value)
+    })
+
+    console.log(url.toString())
```
remove `log`

DiogoSoaress commented on src/logic/safe/store/actions/transactions/gatewayTransactions.ts:
```diff
@@ -1,9 +1,12 @@
 export const ADD_HISTORY_TRANSACTIONS = 'ADD_HISTORY_TRANSACTIONS'
 export const addHistoryTransactions = createAction<HistoryPayload>(ADD_HISTORY_TRANSACTIONS)
 
+export const REMOVE_HISTORY_TRANSACTIONS = 'REMOVE_HISTORY_TRANSACTIONS'
```
Does it make sense to create the analog action for QueuedTransactions?

DiogoSoaress commented on src/routes/safe/components/Transactions/TxList/Filter/utils.ts:
```diff
@@ -34,3 +52,66 @@ export const getFilterHelperText = (value: string, error?: FieldError): TextFiel
+
+// Filter helper
+
+export const isTxFilter = (object: ParsedUrlQuery): object is Partial<FilterForm> => {
```
Did you test what happens if I manually enter one of this query params ? Or if I enter it with invalid values? For the case the url is not generated from the Filter form

iamacook replied:
After our discussion, I've left this as `some`, there only needs to be part of the object for a valid filter to exist. The endpoint-specific filter is created in `getHistoryTxListPage` from the object that this function checks.

I've also removed the hidden fields from the URL.

DiogoSoaress replied:
After checking it together it is no harm to send invalid keys or values to the backend. They will simply be ignored or not applied.

DiogoSoaress reviewed (changes requested):
Left couple of comments but it's looking good!

katspaugh commented:
I’ve been testing this branch, filtering by Incoming (w/o any values in fields) returns outgoing txns as well. 🐛 

Also, UX-wise, there’s no indication of the current filter other than the URL. Or even that any filter is enabled at all.
I think the button should not be green if no filter is active.

And another design thing: I would add a small shadow to the filter popup. All the other popups have one.

iamacook commented:
> I’ve been testing this branch, filtering by Incoming (w/o any values in fields) returns outgoing txns as well. 🐛

Are you sending funds to yourself?

> Also, UX-wise, there’s no indication of the current filter other than the URL. Or even that any filter is enabled at all. I think the button should not be green if no filter is active.
>
> And another design thing: I would add a small shadow to the filter popup. All the other popups have one.

I'll adjust these and get back to you.

katspaugh commented:
> Are you sending funds to yourself?

Ah, true. Never mind then.

katspaugh commented:
Maybe this could be the filtered state indication:

<img width="815" alt="Screenshot 2022-06-08 at 13 01 30" src="https://user-images.githubusercontent.com/381895/172600754-df5769b2-3798-4621-9a5e-ead8a962e255.png">

iamacook commented:
> Maybe this could be the filtered state indication:
> 
> <img alt="Screenshot 2022-06-08 at 13 01 30" width="815" src="https://user-images.githubusercontent.com/381895/172600754-df5769b2-3798-4621-9a5e-ead8a962e255.png">

Altered accordingly:

![image](https://user-images.githubusercontent.com/20442784/172607444-a0725240-4cdf-42b5-b1d7-f8875e7b8aef.png)
![image](https://user-images.githubusercontent.com/20442784/172607480-be7618ee-90be-46ed-8b06-9a8ff63b40b4.png)

katspaugh commented:
Nice!
Is there are more neutral-looking button style? All-grey looks like it's disabled. Not a big deal tho, it is in a way inactive.

katspaugh commented:
Yes, that's more appropriate IMHO.

iamacook commented:
@katspaugh, I've updated it:

![image](https://user-images.githubusercontent.com/20442784/172621892-45035f26-21f0-4e2f-85a0-d409c2c0280f.png)

iamacook commented:
> In v2 we should make the Recipient field auto-complete from the Address Book, and the Token Address – from the hodled tokens.

Both should autocomplete from the address book already. Are you testing on the PR?

I like the idea of the token address from the asset list. We can add it to the list for when we enable date filtering. What do you think? I can, of course, implement it in this version though.

katspaugh commented:
Ah, I don't have an AB on there. Nice touch then!
Tokens we can do with dates, yes. 👍

iamacook commented on src/logic/safe/store/actions/transactions/fetchTransactions/loadGatewayTransactions.ts:
```diff
@@ -1,13 +1,99 @@
+    return url.toString()
+  }
+
+  historyPointers[chainId][safeAddress].next = getPageUrl(txListPage?.next)
+  historyPointers[chainId][safeAddress].previous = getPageUrl(txListPage?.previous)
```
If the fetch throws, `txListPage` will be `undefined`. You can't access properties of `undefined`.

![image](https://user-images.githubusercontent.com/20442784/173011403-02d32f8b-8ee5-4afa-8fb5-b3a664cd20cc.png)

iamacook replied:
Please see my other comment.

iamacook commented on src/logic/safe/store/actions/transactions/fetchTransactions/loadGatewayTransactions.ts:
```diff
@@ -1,13 +1,99 @@
+  safeAddress: string,
+  filter?: FilterForm | Partial<FilterForm>,
+): Promise<TransactionListPage> => {
+  let txListPage: TransactionListPage = {
+    next: undefined,
+    previous: undefined,
+    results: [],
+  }
```
This relates to my other comment. You can't access properties of `undefined`. This is somewhat of a fallback.

iamacook replied:
If the fetch throws then `txListPage` would be `undefined` and you cannot access properties of undefined:

![image](https://user-images.githubusercontent.com/20442784/173011857-f2263cf6-423d-45b1-a1aa-1488ae42d9d4.png)

francovenica commented:
Safe used for testing:
https://pr3870--safereact.review-safe.gnosisdev.com/app/rin:0x9913B9180C20C6b0F21B6480c84422F6ebc4B808

Some minor stuff
If the name is too long in the recipient it reaches the "X". maybe you can cut the name with a "..." before the X.
![image](https://user-images.githubusercontent.com/4503659/173496673-421a7ef0-86fd-423a-9beb-d08b556e6784.png)

The X shows in the recipient to clear the field, but after doing a filter it doesn't show again until you start to write something there or delete a character from the previous filter attempt. This wouldn't be such a hassle if selecting all the text and pressing backspace would not fail to delete what is written (but this is an issue present even in the contract interaction form)

Major issue:
If in the safe given as example you filter by "incoming" and "1" in the amount you get a single result of a spending limit sent to the safe itself
https://pr3870--safereact.review-safe.gnosisdev.com/app/rin:0x9913B9180C20C6b0F21B6480c84422F6ebc4B808/transactions/history?type=Incoming&value=1
![image](https://user-images.githubusercontent.com/4503659/173497244-b1669532-e9ee-418e-b24c-2a54b3e6b140.png)

If you search for only incoming with no amount you get a ton of results, many with "1's" in them: 
https://pr3870--safereact.review-safe.gnosisdev.com/app/rin:0x9913B9180C20C6b0F21B6480c84422F6ebc4B808/transactions/history?type=Incoming
![image](https://user-images.githubusercontent.com/4503659/173497339-e4cf3b42-f67c-47c8-81d1-0c7f570eaf1a.png)
![image](https://user-images.githubusercontent.com/4503659/173497369-5d22a09e-d196-4e04-901d-ac2fba8ab26e.png)

Why so many of those results are ignored when searching with the amount of 1 in the first try?

iamacook commented:
> If the name is too long in the recipient it reaches the "X". maybe you can cut the name with a "..." before the X. ![image](https://user-images.githubusercontent.com/4503659/173496673-421a7ef0-86fd-423a-9beb-d08b556e6784.png)

I will look into this.

> The X shows in the recipient to clear the field, but after doing a filter it doesn't show again until you start to write something there or delete a character from the previous filter attempt. This wouldn't be such a hassle if selecting all the text and pressing backspace would not fail to delete what is written (but this is an issue present even in the contract interaction form)

As it is "an issue present even in the contract interaction form", I would argue that this is out of scope. MUI adds the button as it is related to current user interaction so I personally think it's intended behaviour.

> Major issue: If in the safe given as example you filter by "incoming" and "1" in the amount you get a single result of a spending limit sent to the safe itself https://pr3870--safereact.review-safe.gnosisdev.com/app/rin:0x9913B9180C20C6b0F21B6480c84422F6ebc4B808/transactions/history?type=Incoming&value=1 ![image](https://user-images.githubusercontent.com/4503659/173497244-b1669532-e9ee-418e-b24c-2a54b3e6b140.png)

If you send funds to yourself, you are effectively creating an icoming and outgoing transaction. Afair, the gateway shows these only as incoming transactions. This was discussed with @johannesmoormann and, as it is a backend-related issue and an edge case (users won't be sending funds to themselves), it is fine.

> If you search for only incoming with no amount you get a ton of results, many with "1's" in them: https://pr3870--safereact.review-safe.gnosisdev.com/app/rin:0x9913B9180C20C6b0F21B6480c84422F6ebc4B808/transactions/history?type=Incoming ![image](https://user-images.githubusercontent.com/4503659/173497339-e4cf3b42-f67c-47c8-81d1-0c7f570eaf1a.png) ![image](https://user-images.githubusercontent.com/4503659/173497369-5d22a09e-d196-4e04-901d-ac2fba8ab26e.png)
> 
> Why so many of those results are ignored when searching with the amount of 1 in the first try?

Looking at the network tab, it seems like it's contacting the wrong endpoint. I will look into this as well.

## Findings to tackle
- [ ] Truncate field text if passing the clear button
- [ ] Fix filter endpoint being used

iamacook commented:
@francovenica, the `value` is now converted to wei which mirrors what you see in the transaction list. It should seem 'correct' now from the user's perspective.

I've also adjusted the clear button to always show when there is a value in the field and the text truncates if it goes beyond the button, as seen here:

![address input](https://user-images.githubusercontent.com/20442784/173536858-a53575fc-6dd1-4197-adcc-0fbb742d84d7.gif)

usame-algan reviewed (approved):
Great feature, good work on this! 🚀

francovenica commented:
The issues reported (the X for the clear and searching for an amount of 1) where fixed. Nice

Issues/Questions:

Suggestion: We should probably think about a range for the amount input; if somebody sent you 123 USDC, unless you remember that exact number you are not going to find the tx
Probably something we can do in a later version of the filter.

Question/Suggestion: The "recipient" input for an incoming type of filter is skewing the result, since the only recipient for incoming tx is the safe itself, so the only value that it works is the safe address.
Maybe we can grey out that input if the user chooses the incoming type of filter

We still have the issue of the "outgoing tx" filter only searching for the base token (ETH in rinkeby). I mentioned this in an older comment. Do we still need input from people from the backend about this?

iamacook commented:
Just a follow up: I misunderstood the searching of tokens for *outgoing* transactions. It's *incoming* that has the `tokenAddress` filter.

I cannot see a query for it on the `outgoing-transaction` swagger. This is something that needs to be discussed with backend as well. cc @fmrsabino

fmrsabino commented:
@iamacook The issue with the date filters has been resolved and it is live on `staging`
